src/main/java/org/apache/ibatis/annotations/SelectProvider.java

```diff
@@ -31,4 +34,10 @@
   Class<?> type();
 
   String method();
+
+  /**
+   * Specify your own implementation
+   * @since 3.4.6
+   */
+  Class<? extends AbstractProviderSqlSource> providerSqlSource() default ProviderSqlSource.class;
```

Comment:
I like the spirit of this change but if you are doing this, you should just create an interface for the ProviderSqlSource.
ProviderSqlSource is only reference in 3 areas so should be a pretty easy straightforward change. If you are concerned about changing other classes then naming it IProviderSqlSource in C++ style is fine (although as with most java projects we don't have a 'standard' interface denotation).
We should also just pull a provider interface for these as all the sub provider interfaces have a lot of similarities.
As @jeffgbutler mentioned, I don't think we support java 8 code style yet.
Although, I need to verify this as our travis builds seem to be using java8 to compile everything.

Reply:
use of `ISomeInterface` breaks with Polymorphism - you're exposing an implementation detail. If you ever want to change it from an interface to an abstract class you would need rename the `ISomeInterface` to `ASomeAbstractClass` or something like it - then everything would break.

I definitely think `ProviderSqlSource` should be an interface. Rename the current implementation to `DefaultProviderSqlSource`

Reply:
@h3adache It is more appropriate to use the interface here.  

I agree with @chb0github definition of interface name.
 
After changing to interface , the method of using the constructor to initialize is not friendly enough.  
 
Can I define an init method for initialization, and provides an abstract class for the initialization method?